File: app/src/main/java/com/yawar/memo/repositry/BlockUserRepo.java
package com.yawar.memo.repositry;

import android.annotation.SuppressLint;
import android.app.Application;
import android.util.Log;

import androidx.lifecycle.MutableLiveData;

import com.android.volley.Request;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.StringRequest;
import com.yawar.memo.constant.AllConstants;
import com.yawar.memo.model.ChatRoomModel;
import com.yawar.memo.model.UserModel;
import com.yawar.memo.retrofit.RetrofitClient;
import com.yawar.memo.utils.BaseApp;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;


    public class BlockUserRepo {
        public final MutableLiveData<ArrayList<UserModel>> userBlockListMutableLiveData;
        public   ArrayList<UserModel> userBlockList;
        public MutableLiveData<String> blockedForRepo;
        public MutableLiveData<Boolean> blockedRepo;
        public MutableLiveData<Boolean> unBlockedRepo;
        BaseApp myBase = BaseApp.getInstance();
        ChatRoomRepo chatRoomRepo = myBase.getChatRoomRepo();






        public BlockUserRepo(Application application) { //application is subclass of context

            //cant call abstract func but since instance is there we can do this
            userBlockList = new ArrayList<>();
            userBlockListMutableLiveData = new MutableLiveData<>();
            blockedForRepo = new MutableLiveData<>(null);
            blockedRepo = new MutableLiveData<>(null);
            unBlockedRepo = new MutableLiveData<>(null);

        }
        public MutableLiveData<String> getBlockedForRepo() {
            return blockedForRepo;
        }

        public void setBlockedForRepo(String blockedForRepo) {
            this.blockedForRepo.setValue(blockedForRepo);
        }

        public MutableLiveData<Boolean> getBlockedRepo() {
            return blockedRepo;
        }

        public void setBlockedRepo(Boolean blockedRepo) {
            this.blockedRepo.setValue(blockedRepo);
        }

        public MutableLiveData<Boolean> getUnBlockedRepo() {
            return unBlockedRepo;
        }

        public void setUnBlockedRepo(Boolean blockedRepo) {
            this.unBlockedRepo.setValue(blockedRepo);
        }

        @SuppressLint("CheckResult")
        public void getUserBlock(String user_id) {
            userBlockList= new ArrayList<>();
            userBlockList.clear();


            Single<String> observable = RetrofitClient.getInstance(AllConstants.base_url).getapi().getBlockKist(user_id)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread());
            observable.subscribe(s -> {
                try {


                    System.out.println(s + "responeeeeeeee");
                    JSONObject respObj = new JSONObject(s);
                    JSONArray jsonArray = (JSONArray) respObj.get("data");

                    for (int i = 0; i <= jsonArray.length() - 1; i++) {
                        JSONObject jsonObject = jsonArray.getJSONObject(i);
                        //System.out.println(jsonObject.getString("last_message"));
                        String image = jsonObject.getString("profile_image");

                        String special_number = jsonObject.getString("sn");
                        String fName = jsonObject.getString("first_name");
                        String lName = jsonObject.getString("last_name");
                        String phone = jsonObject.getString("phone");
                        String userId = jsonObject.getString("id");
                        String email = jsonObject.getString("email");
                        String blockedFor = jsonObject.getString("blocked_for");

                        userBlockList.add(new UserModel(userId, fName, lName, email, phone, special_number, image, blockedFor));
                    }

                    userBlockListMutableLiveData.setValue(userBlockList);
                }catch (Exception e){
                    Log.d("Error", "getUserBlock: "+e);
                }





                    },
                    s -> {
                        userBlockList = null;
                        userBlockListMutableLiveData.setValue(null);
                    });
        }
        public MutableLiveData<ArrayList<UserModel>> getUserBlockList() {

            return userBlockListMutableLiveData;
        }
        public  void deleteBlockUser(String user_id,String status){
            for(UserModel user:userBlockList){
                if(user.getUserId().equals(user_id)){
                    user.setStatus(status);
                    break;
                }}
            userBlockListMutableLiveData.postValue(userBlockList);
        }
        public  void addBlockUser(UserModel userModel){
            boolean searchBlock = false;
            for(UserModel user:userBlockList){
                if(user.getUserId().equals(userModel.getUserId())){
                    user.setStatus(userModel.getStatus());
                    searchBlock = true;
                    break;
                }}
            if(!searchBlock){
            userBlockList.add( 0,userModel);}
            userBlockListMutableLiveData.postValue(userBlockList);
        }

        @SuppressLint("CheckResult")
        public void sendBlockRequest(String my_id, String anthor_user_id) {


            Single<String> observable = RetrofitClient.getInstance(AllConstants.base_node_url).getapi().blockUser(my_id,anthor_user_id)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread());
            observable.subscribe(s -> {
                        try {
                    String blokedForRespone = "";
                    boolean blockedRespone = false;
                    try {
                        System.out.println("blockedddddddd");
                        JSONObject jsonObject = new JSONObject(s);
                        blokedForRespone = jsonObject.getString("blocked_for");
                        blockedRespone= jsonObject.getBoolean("blocked");


                    } catch (JSONException e) {
                        e.printStackTrace();
                    }
                    blockedForRepo.setValue(blokedForRespone);

                    chatRoomRepo.setBlockedState(anthor_user_id,blokedForRespone);
                    blockedRepo.setValue(true);

//


                    }catch (Exception e){
                        }
                    },
                    s -> {
                       System.out.println("problem");
                    });

        }
        @SuppressLint("CheckResult")
        public void sendUnbBlockUser(String my_id, String anthor_user_id) {
            Single<String> observable = RetrofitClient.getInstance(AllConstants.base_node_url).getapi().unBlockUser(my_id,anthor_user_id)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread());


            observable.subscribe(s -> {
                        try {
                            System.out.println("unnnnBlockeddddddd");

                            String blokedForRespone = "";
                            Boolean unBlockedRespone = false;

                            try {
                                JSONObject jsonObject = new JSONObject(s);
                                blokedForRespone = jsonObject.getString("blocked_for");

                            } catch (JSONException e) {
                                e.printStackTrace();
                            }
                            blockedForRepo.setValue(blokedForRespone);
                            unBlockedRepo.setValue(unBlockedRespone);
                            chatRoomRepo.setBlockedState(anthor_user_id, blokedForRespone);
//


                        } catch (Exception e) {

                        }
                    },
                    s -> {
                        System.out.println("problem");
                    });



        }
    }

File: app/src/main/java/com/yawar/memo/repositry/BlockUserRepo.kt
package com.yawar.memo.repositry

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.yawar.memo.Api.ChatApi
import com.yawar.memo.BaseApp
//import com.yawar.memo.Api.GdgApi
import com.yawar.memo.constant.AllConstants
import com.yawar.memo.domain.model.UserModel
//import com.yawar.memo.repositry.chatRoomRepo.ChatRoomRepoImp
import dagger.hilt.android.AndroidEntryPoint
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.util.*
import javax.inject.Inject


//  class BlockUserRepo @Inject constructor(private val chatRoomRepoo: ChatRoomRepoImp) {
class BlockUserRepo  @Inject constructor(private val chatApi: ChatApi,
                                         private val chatRoomRepoo: ChatRoomRepoo) {

    var myBase = BaseApp.instance
//    var chatRoomRepoo = myBase.chatRoomRepoo

    private val _userBlockListMutableLiveData = MutableLiveData<ArrayList<UserModel?>?>()
    val userBlockListMutableLiveData: LiveData<ArrayList<UserModel?>?>
        get() = _userBlockListMutableLiveData


    private val _blockedForRepo = MutableLiveData<String>()
    val blockedForRepo: LiveData<String>
        get() = _blockedForRepo


    private val _blockedRepo =  MutableLiveData<Boolean>()
    val isBlocked : LiveData<Boolean>
        get() = _blockedRepo



      private val _unBlockedRepo =  MutableLiveData<Boolean>()
    val isUnBlocked  : LiveData<Boolean>
        get() = _unBlockedRepo


      private var viewModelJob = Job()
    private val coroutineScope = CoroutineScope(viewModelJob + Dispatchers.Main )

     fun getUserBlock( user_id: String) {
         coroutineScope.launch {
             val userBlockList = ArrayList<UserModel?>()

//
//             var getBlockDeferred = GdgApi(AllConstants.base_node_url).apiService
//
//                 .getBlockKist(user_id)
//             var getBlockDeferred = GdgApi.apiService
//
//                 .getBlockKist(user_id)
             var getBlockDeferred = chatApi

                 .getBlockKist(user_id)


             try {
                 var listResult = getBlockDeferred?.await()

                 val jsonArray = JSONArray(listResult) as JSONArray
                 Log.d("getUserBlock", jsonArray.length().toString())
                 for (i in 0 .. jsonArray.length()-1) {

                     val jsonObject = jsonArray.getJSONObject(i)

                     val image = jsonObject.getString("profile_image")
                     val special_number = jsonObject.getString("sn")
                     val fName = jsonObject.getString("first_name")
                     val lName = jsonObject.getString("last_name")
                     val phone = jsonObject.getString("phone")
                     val userId = jsonObject.getString("id")
                     val email = jsonObject.getString("email")

                     userBlockList.add(
                         UserModel(userId,
                         fName,
                         lName,
                         email,
                         phone,
                         special_number,
                         image,
                         "null")
                     )


                 }
                 Log.d("_userBlockListMutableLiveData.value", userBlockList.size.toString())


                 _userBlockListMutableLiveData.value = userBlockList


             } catch (e: Exception) {


                 Log.d("getMarsRealEstateProperties: ","Failure: ${e.message}")

             }
         }
     }

    fun addBlockUser(userModel: UserModel) {
        var searchBlock = false
        var  userBlockList = _userBlockListMutableLiveData.value

        if (userBlockList != null) {
            for (user in userBlockList) {
                if (user != null) {
                    if (user.userId == userModel.userId) {
                        user.status = userModel.status
                        searchBlock = true
                        break
                    }
                }
            }
        }
        if (!searchBlock) {
            if (userBlockList != null) {
                userBlockList.add(0, userModel)
            }
        }
        _userBlockListMutableLiveData.value = userBlockList
    }


    fun sendBlockRequest( my_id: String,  anthor_user_id:String) {
        coroutineScope.launch {
            Log.d("sendBlockRequest", "sendBlockRequest: ")
//
//            val respone =   GdgApi(AllConstants.base_node_url).apiService.blockUser(my_id, anthor_user_id)

//            val respone =   GdgApi.apiService.blockUser(my_id, anthor_user_id)
            val respone =   chatApi.blockUser(my_id, anthor_user_id)




            try {
                val listResult = respone?.await()

                var blokedForRespone = ""
                var blockedRespone = false

                val jsonObject =   JSONObject(listResult.toString())
                blokedForRespone = jsonObject.getString("blocked_for")
                blockedRespone = jsonObject.getBoolean("blocked")
                _blockedForRepo.value = blokedForRespone
                Log.d("sendBlockUser", blokedForRespone)


                chatRoomRepoo.setBlockedState(anthor_user_id, blokedForRespone)
                _blockedRepo.value = true

            } catch (e: Exception) {


                Log.d("getMarsRealEstateProperties: ", "Failure: ${e.message}")

            }

        }

    }


    fun sendUnbBlockUser( my_id: String,  anthor_user_id:String) {
        coroutineScope.launch {

            Log.d("sendUnbBlockUser", "sendUnbBlockUser: ")

//            var respone = GdgApi(AllConstants.base_node_url).apiService
//                .unBlockUser(my_id, anthor_user_id)
//            var respone = GdgApi.apiService
//                .unBlockUser(my_id, anthor_user_id)
            var respone = chatApi
                .unBlockUser(my_id, anthor_user_id)
            try {
                val listResult = respone?.await()

                var blokedForRespone = ""
                val unBlockedRespone = false

                try {

                    val jsonObject = JSONObject(listResult)
                    blokedForRespone = jsonObject.getString("blocked_for")
                } catch (e: JSONException) {
                    e.printStackTrace()
                }

                _blockedForRepo.value = blokedForRespone
                _unBlockedRepo.value = true
                Log.d("sendUnbBlockUser", blokedForRespone)


                chatRoomRepoo.setBlockedState(anthor_user_id, blokedForRespone)
                deleteBlockUser(anthor_user_id)


            } catch (e: Exception) {


                Log.d("getMarsRealEstateProperties: ", "Failure: ${e.message}")

            }

        }

    }
    fun deleteBlockUser(user_id: String) {
        val userBlockList = _userBlockListMutableLiveData.value

        if (userBlockList != null) {
            for (user in userBlockList) {
                if (user != null) {
                    if (user.userId == user_id) {
                        //                    user.setStatus(status);
                        userBlockList.remove(user)
                        break
                    }
                }
            }
        }
        _userBlockListMutableLiveData.value = userBlockList
    }

    fun setBlockedForRepo(blockedForRepo: String) {
        _blockedForRepo.value = blockedForRepo
    }
    fun setBlockedRepo(blockedRepo: Boolean) {
        _blockedRepo.value = blockedRepo
    }



    fun setUnBlockedRepo(blockedRepo: Boolean) {
        _unBlockedRepo.value = blockedRepo
    }
      fun setUserBlockListMutableLiveData(data : ArrayList<UserModel?>?) {
          _userBlockListMutableLiveData.value = data
      }

}